### ISSUE TITLE
The reversed mask of the targetted label can't converge on MNIST dataset

### ISSUE THREAD

**Issue by Crane-Mocker** [bolunwang/backdoor#12]
My classmate and I changed the parameters in `visualizer.py` and `gtsrb_visualize_example.py`, which is `MNIST_visualize_example.py` in our case, to reproduce the reverse engineering process on MNIST dataset. When it started to reverse the triggers, the mask of the targeted label cannot converge, so that it cannot tell the properly targeted label. 

Here is the result of  mad outlier detection. The targetted label is 5. The trigger is a little white square on the right bottom.
![2021-05-27 21-22-14 的屏幕截图](https://user-images.githubusercontent.com/46662246/119834018-28e07200-bf32-11eb-92ca-e7abe70a0e25.png)

Here are the results of the reversed masks of each label.
![res](https://user-images.githubusercontent.com/46662246/119834860-e703fb80-bf32-11eb-9d2a-520ea10e9d39.png)

Here are the results of  the reversed masks of label 5 in each step.
![process](https://user-images.githubusercontent.com/46662246/119835140-24688900-bf33-11eb-8071-dcb76f7d7732.png)

Here are our parameters in `MNIST_visualize_example.py` 

```python
# # input size
IMG_ROWS = 28
IMG_COLS = 28
IMG_COLOR = 1
IMAGE_SHAPE = (IMG_ROWS, IMG_COLS, IMG_COLOR)
mad_outlier_detection
#
CLASSES_ALL = 10  # total number of classes in the model
Y_TARGET = 5  # (optional) infected target label, used for prioritizing label scanning

PREPROCESS = 'mnist'  # preprocessing method for the task, GTSRB uses raw pixel intensities

# parameters for optimization
BATCH_SIZE = 32  # batch size used for optimization
LR = 0.1  # learning rate
STEPS = 1000  # total optimization iterations
NB_SAMPLE = 1000  # number of samples in each mini batch
MINI_BATCH = NB_SAMPLE // BATCH_SIZE  # mini batch size used for early stop
INIT_COST = 2e-3  # initial weight used for balancing two objectives
```
We have tried to adjust the other parameters, too. And we found that when `LR` was smaller and `INIT_COST` was a bit larger, like `2e-3`, the result would be a bit closer to the proper targetted label .

**Comment by Crane-Mocker** [comment]
We solved this problem. At first we used LeNet5 as follow

```python3
input_shape = (28, 28, 1)
model = Sequential()
model.add(Conv2D(input_shape=input_shape, kernel_size=(5, 5), filters=20, activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2), strides=2, padding='same'))

model.add(Conv2D(kernel_size=(5, 5), filters=50, activation='relu', padding='same')) model.add(MaxPooling2D(pool_size=(2, 2), strides=2, padding='same'))
model.add(Flatten())
model.add(Dense(500, activation='relu'))
model.add(Dense(10, activation='softmax'))
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
opt = keras.optimizers.adam(lr=0.001, decay=1 * 10e-5)
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])
```
Then we changed it as this
```python
model = Sequential()
model.add(Conv2D(filters=6, kernel_size=(5, 5), padding='valid', input_shape=(28, 28, 1), activation='tanh'))  # C1
model.add(MaxPooling2D(pool_size=(2, 2)))    # S2
model.add(Conv2D(filters=16, kernel_size=(5, 5), padding='valid', activation='tanh'))   # C3
model.add(MaxPooling2D(pool_size=(2, 2)))    # S4
model.add(Flatten())
model.add(Dense(120, activation='tanh'))     # C5
model.add(Dense(84, activation='tanh'))      # F6
model.add(Dense(10, activation='softmax'))  # output
# sgd = SGD(lr=0.05, decay=1e-6, momentum=0.9, nesterov=True)
model.summary()
opt = keras.optimizers.adam(lr=0.001, decay=1 * 10e-5)
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])
```
And we got the right result.
